spring-integration-core/src/main/java/org/springframework/integration/handler/BarrierMessageHandler.java

```
@@ -0,0 +1,192 @@
+			}
+		}
+		catch (InterruptedException e) {
+			Thread.currentThread().interrupt();
+			throw new MessagingException(requestMessage, "Interrupted while waiting for release", e);
+		}
+		finally {
+			this.inProcess.remove(key);
+			this.suspensions.remove(key);
+		}
+		return null;
+	}
+
+	private Object processRelease(Message<?> requestMessage, Object key, Message<?> release) {
+		this.suspensions.remove(key);
+		if (release.getPayload() instanceof Throwable) {
+			throw new MessagingException(requestMessage, "Releasing flow returned a throwable",
+					(Throwable) release.getPayload());
+		}
+		else {
+			return buildResult(key, requestMessage, release);
+		}
+	}
+
+	/**
+	 * Override to change the default mechanism by which the inbound and release messages
+	 * are returned as a result.
+	 * @param key The correlation key.
+	 * @param requestMessage the inbound message.
+	 * @param release the release message.
+	 * @return the result.
+	 */
+	protected Object buildResult(Object key, Message<?> requestMessage, Message<?> release) {
+		SimpleMessageGroup group = new SimpleMessageGroup(key);
+		group.add(requestMessage);
+		group.add(release);
+		return this.messageGroupProcessor.processMessageGroup(group);
+	}
+
+	private SynchronousQueue<Message<?>> createOrObtainQueue(Object key) {
+		SynchronousQueue<Message<?>> syncQueue = new SynchronousQueue<Message<?>>();
+		SynchronousQueue<Message<?>> existing = this.suspensions.putIfAbsent(key, syncQueue);
+		if (existing != null) {
+			syncQueue = existing;
+		}
+		return syncQueue;
+	}
+
+	@Override
+	public void triggerAction(Message<?> message) {
+		Object key = this.correlationStrategy.getCorrelationKey(message);
+		if (key == null) {
+			throw new MessagingException(message, "Correlation Strategy returned null");
+		}
+		SynchronousQueue<Message<?>> syncQueue = createOrObtainQueue(key);
+		try {
+			if (syncQueue != null && !syncQueue.offer(message, timeout, TimeUnit.MILLISECONDS)) {
```

Actually - I don't think this is a good idea after all.

If we allow concurrent use of the same correlation id, it is indeterminate as to which trigger will arrive first.

I think it's safer to fail fast.